File: algorithm/leetcode/q0268/main.cpp
#include<vector>
using namespace std;

class Solution {
public:
    int missingNumber(vector<int>& nums) {
        auto n = static_cast<int>(nums.size());
        while (true) {
            auto exchaned = false;
            for (auto i=0; i<n; i++) {
                auto i_holder = nums[i];
                while (i_holder != i && i_holder != n) {
                    exchaned = true;
                    auto temp = nums[i_holder];
                    nums[i_holder] = i_holder;
                    i_holder = temp;
                }
                nums[i] = i_holder;
            }
            if (!exchaned) {
                for (auto i=0; i<n; i++) {
                    if (i!=nums[i]) {
                        return i;
                    }
                }
                return n;
            }
        }
    }
};

#include<cassert>

int main() {
    auto s = Solution();
    auto nums = vector<int> {0, 1};
    assert(s.missingNumber(nums) == 2);
}

File: algorithm/leetcode/q0268/main.py
class Solution(object):
    def missingNumber(self, nums: list[int]):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        while True:
            exchanged = False
            missing = n
            for i in range(n):
                i_holder = nums[i]
                while i != i_holder and i_holder != n:
                    exchanged = True

                    # 需要放到目标位置 nums[i_holder]
                    temp = nums[i_holder]
                    nums[i_holder] = i_holder
                    i_holder = temp
                # 最后归宿 i
                nums[i] = i_holder
                if i != i_holder:
                    if exchanged and i_holder != n:
                        return i
                    missing = i
                
            if not exchanged:
                return missing

if __name__ == '__main__':
    m = Solution().missingNumber
    assert m([0]) == 1
    assert m([1]) == 0, m([0])
    assert m([0, 1]) == 2
    assert m([1, 0]) == 2
    assert m([3, 0, 1]) == 2, m([3, 0, 1])
    assert m([9,6,4,2,3,5,7,0,1]) == 8

